~SPIRVModuleImpl() crashes when delete no-id entries

lifpan opened the pull request:
It crashes because it tries to delete some no-id entry (OpStore) twice, which is
pushed into vector "EntryNoId" twice.

Do not add entries of SPIRVBasicBlock and SPIRVInstruction into Module when call
Decoder.getEntry(). They will be added at addBasicBlock() and addInstruction().

Commit:
~SPIRVModuleImpl() crashes when delete no-id entries

It crashes because it tries to delete some no-id entry (OpStore) twice, which is
pushed into vector "EntryNoId" twice.

Do not add entries of SPIRVBasicBlock and SPIRVInstruction into Module when call
Decoder.getEntry(). They will be added at addBasicBlock() and addInstruction().

bader commented:
Thanks for the patch, but it looks like it doesn't resolve the issue in all cases.
Please, take a look at the Trevis results. There are failing tests, which crash in ~SPIRVModuleImpl().

lifpan commented:
Where to get the test and how to run it at my local?

bader commented:
@lifpan, tests are located in test/SPIRV/ directory. 
There are several ways to run these tests, but easiest way to validate your patch is to build 'check-all' target e.g. 'make check-all'.

yxsamliu commented:
you can also use llvm-lit -v followed by the SPIR-V test directory, which should be at llvm/test/SPIRV.

Commit:
~SPIRVModuleImpl() crashes when delete no-id entries

No-id entries of OpGroupDecorate and OpGroupMemberDecorate are pushed into vector
"EntryNoId" twice. See SPIRVGroupDecorateGeneric::decode(std::istream &I).
No-id entry of OpLine will be deleted by std::shared_ptr automatically.

lifpan commented:
Fix code for OpGroupDecorate, OpLine, and OpUndef.
Pass llvm-lit -v test/SPIRV:
  Expected Passes    : 111
  Expected Failures  : 1

AlexeySotkin commented on lib/SPIRV/libSPIRV/SPIRVModule.cpp:
```diff
@@ -552,7 +553,8 @@ SPIRVModuleImpl::addEntry(SPIRVEntry *Entry) {
       IdEntryMap[Id] = Entry;
   } else {
-    EntryNoId.push_back(Entry);
+    if (EntryNoId.empty() || EntryNoId.back() != Entry)
```
Why this condition? Probably we can use std::set to avoid duplicates.

lifpan replied:
For example, entry of OpGroupDecorate will be added twice when decode SPIRV binary:
1) SPIRVGroupDecorateGeneric::decode(std::istream &I) -> Module->addGroupDecorateGeneric(this)
2) Decoder.getEntry() -> M.add(Entry)

AlexeySotkin replied:
As far as  I understand with this condition the entry will be added to the vector if 1) the vector is empty or 2) if the last element of the vector equals to the entry. What if the entry is already in the vector but it is not at the end of the vector?

Commit:
~SPIRVModuleImpl() crashes when delete no-id entries

Use type std::set for "EntryNoId" to avoid duplicate elements.The elements
in std:set are sorted by memory pointer, not by creation order. An entry of
OpLine (A) can be deleted because of deletion of another entry (B) which refers
to A. If A is placed after B, it will try to access invalid A when it frees
"EntryNoId". Therefore, not insert entry of OpLine to "EntryNoId", and it will
be deleted by std::shared_ptr automatically.

yxsamliu commented:
LGTM. Thanks.

lifpan commented:
@yxsamliu I found a mistake near line 1508 in SPIRVModule.cpp, that Decoder.getEntry() may return NULL. This can happen if the binary is not strictly encoded.

Please help correct it:
   while(Decoder.getWordCountAndOpCode())
     M.add(Decoder.getEntry());
-->
  while (Decoder.getWordCountAndOpCode()) {
    SPIRVEntry *Entry = Decoder.getEntry();
    if (Entry != NULL)
      M.add(Entry);
  }

bader commented:
@lifpan, could you create another pull request and attach the test case, which reproduce the issue, please?
